packages/service-library/tests/fastapi/test_request_decorators.py

```diff
@@ -0,0 +1,248 @@
+from servicelib.fastapi.requests_decorators import TASK_NAME_PREFIX, cancellable_request
+from tenacity._asyncio import AsyncRetrying
+from tenacity.stop import stop_after_attempt
+from tenacity.wait import wait_fixed
+
+
+@pytest.fixture
+def app() -> FastAPI:
+
+    api_router = APIRouter()
+
+    @api_router.get("/")
+    def _get_root():
+        return {"name": __name__, "timestamp": datetime.utcnow().isoformat()}
+
+    @api_router.get("/sleep/{delay}")
+    @cancellable_request
+    async def _cancellable_with_await(_request: Request, delay: NonNegativeInt):
+        await asyncio.sleep(delay)
+
+    # TODO: handler spawns subprocesses
+    # TODO: handler spawns threads
+    # TODO: handler spawns aio-tasks
+    # TODO: handler with backgroundtask
+
+    _app = FastAPI()
+    _app.include_router(api_router)
+
+    @_app.on_event("startup")
+    async def on_startup():
+        _app.state.started = True
+        _app.state.stopped = False
+
+    @_app.on_event("shutdown")
+    async def on_shutdown():
+        _app.state.started = True
+        _app.state.stopped = True
+
+    return _app
+
+
+@pytest.fixture
+async def test_client(app: FastAPI) -> AsyncIterable[TestClient]:
+    async with TestClient(app) as client:
+        yield client
+
+
+@pytest.fixture()
+def inspect_app_tasks(
+    event_loop: AbstractEventLoop,
+) -> Callable[[], list[Task]]:
+    """Function to return cancellable tasks"""
+
+    def go():
+        return [
+            task
+            for task in asyncio.all_tasks(event_loop)
+            if task.get_name().startswith(f"{TASK_NAME_PREFIX}/")
+        ]
+
+    return go
+
+
+# TESTS
+
+
+def test_raises_if_wrong_signature_upon_import(app: FastAPI):
+    # let's define some routes
+    @app.get("/correct")
+    @cancellable_request
+    async def _correct_signature(_request: Request, x: int):
+        ...
+
+    with pytest.raises(ValueError) as exc_info:
+
+        @app.get("/wrong")
+        @cancellable_request
+        async def _wrong_signature():
+            ...
+
+    assert "_wrong_signature" in str(exc_info.value)
+
+
+async def test_with_non_cancellable_entrypoint(
+    test_client: TestClient, inspect_app_tasks: Callable[[], list[Task]]
+):
+    assert not inspect_app_tasks()
+
+    # NOT cancellable entrypoint
+    r = await test_client.get("/")
+    assert r.status_code == status.HTTP_200_OK
+
+    assert not inspect_app_tasks()
+
+
+async def test_cancellable_handler_with_successful_request(
+    test_client: TestClient, inspect_app_tasks: Callable[[], list[Task]]
+):
+    # cancellable entrypoint
+    r = await test_client.get("/sleep/0")
+    assert r.status_code == status.HTTP_200_OK
+
+    assert all(
+        t.done() for t in inspect_app_tasks()
+    ), "All tasks (if any) should be done"
+
+
+#
+# NOTE:
+# - Cannot find a way to disconnect a client using the TestClient fixture  :-(
+# - Any ideas welcome!
+#
+
+
+@pytest.mark.skip(reason="DEV: cannot emulate disconnect")
```

Comment:
Something a bit more convolute than SAN's suggestion, which I would try out first:
- define a request which takes more that 1 second to reply
- create the client in a separate process.
- Then you can use one of the following:
    - make a request with 0.1 seconds timeout.
    - when starting the request create a thread which after sleeping 0.1 seconds, either kills the request by cancelling it or kills the process itself.

I think that separating the client in a different process should allow you to achieve this.

Reply:
@GitHK I tried something similar but ...  what is the url I connect the client to? How do I start the server?